Make GetTokenForAppAsync less confusing and allow to pass tenantId #413

jmprieur opened the pull request:
Checked with @hpsin and here is what we agreed to:
- Change the signature of `GetAccessTokenForUserAsync` to take a `string` (instead of a `IEnumerable<string>`) as there is only one possible string for a given resource of App Id URI AppIdUri: "AppIdUri/.default". Check that the resource ends in "./default"
- Add an additional optional parameter `tenant` to support this scenario, and verify that this tenant is not organizations (and of course common and consumers, which don't make sense)

```CSharp
public async Task<string> GetAccessTokenForAppAsync(string scope, string? tenant = null)
```

jennyf19 commented on src/Microsoft.Identity.Web/ITokenAcquisition.cs:
```diff
@@ -37,12 +37,15 @@ public interface ITokenAcquisition
         /// using the client credentials flow. See https://aka.ms/msal-net-client-credentials.
         /// </summary>
-        /// <param name="scopes">scopes requested to access a protected API. For this flow (client credentials), the scopes
+        /// <param name="scope">scope requested to access a protected API. For this flow (client credentials), the scope
```
```suggestion
        /// <param name="scope">The scope requested to access a protected API. For this flow (client credentials), the scope
```

jennyf19 commented on src/Microsoft.Identity.Web/ITokenAcquisition.cs:
```diff
@@ -37,12 +37,15 @@ public interface ITokenAcquisition
         /// should be of the form "{ResourceIdUri/.default}" for instance <c>https://management.azure.net/.default</c> or, for Microsoft
         /// Graph, <c>https://graph.microsoft.com/.default</c> as the requested scopes are defined statically with the application registration
-        /// in the portal, and cannot be overridden in the application.</param>
+        /// in the portal, cannot be overridden in the application, and you can request a token for only one resource at a time (use
```
```suggestion
        /// in the portal, cannot be overridden in the application, as you can request a token for only one resource at a time (use
```

jennyf19 commented on src/Microsoft.Identity.Web/ITokenAcquisition.cs:
```diff
@@ -37,12 +37,15 @@ public interface ITokenAcquisition
         /// Graph, <c>https://graph.microsoft.com/.default</c> as the requested scopes are defined statically with the application registration
-        /// in the portal, and cannot be overridden in the application.</param>
+        /// in the portal, cannot be overridden in the application, and you can request a token for only one resource at a time (use
+        /// several call to get tokens for other resources).</param>
```
```suggestion
        /// several calls to get tokens for other resources).</param>
```

jennyf19 commented on src/Microsoft.Identity.Web/ITokenAcquisition.cs:
```diff
@@ -37,12 +37,15 @@ public interface ITokenAcquisition
+        /// in the portal, cannot be overridden in the application, and you can request a token for only one resource at a time (use
+        /// several call to get tokens for other resources).</param>
+        /// <param name="tenant">Enables overriding of the tenant/account for the same identity. This is useful in the
+        /// cases where a given account is guest in other tenants, and you want to acquire tokens for a specific tenant, like where the user is a guest in.</param>
```
```suggestion
        /// cases where a given account is a guest in other tenants, and you want to acquire tokens for a specific tenant.</param>
```

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -81,6 +81,16 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
+        /// meta-tenant identifiers which are not allowed in client credentials.
+        /// </summary>
+        private readonly string[] unwantedTenantIdentifiersInClientCredentials = new string[]
+        {
```
nit: start w/`_` how about `_metaTenantIdentifiers` or `_invalidTenantIdentifiersInClientCredentials` or ...?
Also, we have a const file, so I would put these there as const values, we might use them other places and those can be replaced w/the constant value too.

jmprieur replied:
Thanks for the tip.
Addressed!

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -293,25 +303,41 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
         {
-            if (scopes == null)
+            if (scope == null)
             {
```
nit: you can do string.IsNullOrEmpty(scope) now

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -293,25 +303,41 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
+            }
+
+            if (!scope.EndsWith("/.default", true, CultureInfo.InvariantCulture))
+            {
```
nit: do you need the "/" or is ".default" enough?

jmprieur replied:
I think that /.default is probably better as the resource needs to be ended with /

jennyf19 commented on src/Microsoft.Identity.Web/Constants/IDWebErrorMessage.cs:
```diff
@@ -42,6 +42,10 @@ internal static class IDWebErrorMessage
         public const string TokenIsNotJwtToken = "IDW10403: Token is not JWT token. ";
+        public const string ClientCredentialScopeParameterShouldEndInDotDefault =
+       "IDW10404: 'scope' parameter should be of the form 'AppIdUri/.default'.";
+        public const string ClientCredentialTenantShouldBeTenanted =
```
should an aka.ms link be included here w/more info on why this is the case? is it clear enough for developers to know how to fix it based on the error message? i'm not sure.

jmprieur replied:
Thanks for the suggestion. I've added one:  See https://aka.ms/ms-id-web/daemon-scenarios.

jennyf19 commented on tests/Microsoft.Identity.Web.Test.Integration/AcquireTokenForAppIntegrationTests.cs:
```diff
@@ -61,14 +62,30 @@ public async Task GetAccessTokenForApp_ReturnsAccessTokenAsync()
             AssertAppTokenInMemoryCache(TestConstants.ConfidentialClientId, 1);
         }
 
+        [Fact]
```
you can do different arguments and replace the tenant value w/organziations, common, consumers.
```csharp
[Theory]
[InlineData(Constants.Common)]
[InlineData(Constants.Organizations)]
[InlineData(Constants.Consumers)]
public async Task GetAccessTokenForApp_WithMetaTenant(string metaTenant)
{
....
await _tokenAcquisition.GetAccessTokenForAppAsync(TestConstants.s_scopeForApp, metaTenant).ConfigureAwait(false);
....
}

jmprieur replied:
We do already with the test that uses user scopes.

jmprieur replied:
Addressed

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -293,25 +303,41 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
         /// using the client credentials flow. See https://aka.ms/msal-net-client-credentials.
         /// </summary>
-        /// <param name="scopes">scopes requested to access a protected API. For this flow (client credentials), the scopes
+        /// <param name="scope">scope requested to access a protected API. For this flow (client credentials), the scope
```
```suggestion
        /// <param name="scope">The scope requested to access a protected API. For this flow (client credentials), the scope
```

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -293,25 +303,41 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
         /// should be of the form "{ResourceIdUri/.default}" for instance <c>https://management.azure.net/.default</c> or, for Microsoft
         /// Graph, <c>https://graph.microsoft.com/.default</c> as the requested scopes are defined statically with the application registration
-        /// in the portal, and cannot be overridden in the application.</param>
+        /// in the portal, cannot be overridden in the application, and you can request a token for only one resource at a time (use
```
```suggestion
        /// in the portal, cannot be overridden in the application, as you can request a token for only one resource at a time (use
```

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -293,25 +303,41 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
         /// Graph, <c>https://graph.microsoft.com/.default</c> as the requested scopes are defined statically with the application registration
-        /// in the portal, and cannot be overridden in the application.</param>
+        /// in the portal, cannot be overridden in the application, and you can request a token for only one resource at a time (use
+        /// several call to get tokens for other resources).</param>
```
```suggestion
        /// several calls to get tokens for other resources).</param>
```

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -293,25 +303,41 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
+        /// in the portal, cannot be overridden in the application, and you can request a token for only one resource at a time (use
+        /// several call to get tokens for other resources).</param>
+        /// <param name="tenant">Enables overriding of the tenant/account for the same identity. This is useful in the
+        /// cases where a given account is guest in other tenants, and you want to acquire tokens for a specific tenant, like where the user is a guest in.</param>
```
```suggestion
        /// cases where a given account is a guest in other tenants, and you want to acquire tokens for a specific tenant.</param>
```

jmprieur commented:
Thanks @jennyf19 for these suggestions.
I've addressed all your PR feedback: https://github.com/AzureAD/microsoft-identity-web/pull/424/commits/1df2f006ba9f367bf388062654bf29df51ee708b

hpsin commented:
Consumers is allowed, since it is a single tenant.
It's probably worth keeping the old signature around and updating its comments for future compatibility. If we suddenly start supporting multiple scopes (eg more than just /.default) then we'll want this to take multiple strings again.

hpsin commented on src/Microsoft.Identity.Web/Constants/IDWebErrorMessage.cs:
```diff
@@ -42,6 +42,10 @@ internal static class IDWebErrorMessage
+        public const string ClientCredentialScopeParameterShouldEndInDotDefault =
+       "IDW10404: 'scope' parameter should be of the form 'AppIdUri/.default'. See https://aka.ms/ms-id-web/daemon-scenarios.";
+        public const string ClientCredentialTenantShouldBeTenanted =
+       "IDW10405: 'tenant' parameter should be a tenant ID or domain name, not 'common', 'organizations' or 'consumers'. See https://aka.ms/ms-id-web/daemon-scenarios.";
```
Remove consumers.

hpsin commented on src/Microsoft.Identity.Web/ITokenAcquisition.cs:
```diff
@@ -37,12 +37,15 @@ public interface ITokenAcquisition
+        /// in the portal, cannot be overridden in the application, as you can request a token for only one resource at a time (use
+        /// several calls to get tokens for other resources).</param>
+        /// <param name="tenant">Enables overriding of the tenant/account for the same identity. This is useful in the
+        /// cases where a given account is a guest in other tenants, and you want to acquire tokens for a specific tenant.</param>
```
There's no guest scenario for app only tokens, just SP presence in other tenants. Could say "useful for multi tenant apps" or daemons? Not sure that "same identity" will resonate, since we usually don't treat SPs as identities?

hpsin commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -293,25 +305,40 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
+
+            if (!scope.EndsWith("/.default", true, CultureInfo.InvariantCulture))
+            {
+                throw new ArgumentException(IDWebErrorMessage.ClientCredentialScopeParameterShouldEndInDotDefault, nameof(scope));
```
This is good for the singleton scenario. I think that we should retain the original one and not do checks there, though, in an effort to remain OIDC/oauth compliant.

hpsin commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -81,6 +81,18 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
+            {
+                Constants.Common,
+                Constants.Organizations,
+                Constants.Consumers,
```
Remove consumers.

jmprieur replied:
Thansk @hspin. I addressed them all
